source/agora/crypto/Bech32.d

```diff
@@ -0,0 +1,757 @@
+
+    ubyte[] ret;
+    ret.length = 6;
+    for (size_t i = 0; i < 6; ++i)
+        // Convert the 5-bit groups in mod to checksum values.
+        ret[i] = (mod >> (5 * (5 - i))) & 31;
+
+    return ret;
+}
+
+/*******************************************************************************
+
+    Verify a checksum
+
+    Params:
+        hrp = human redable part
+        values = values to verify the checksum with
+
+    Returns:
+        encoding if the checksum is valid; invalid encoding constant if not
+
+*******************************************************************************/
+
+Encoding verifyChecksum (in char[] hrp, in ubyte[] values) pure nothrow @safe
+{
+    // PolyMod computes what value to xor into the final values to make the
+    // checksum 0. However, if we required that the checksum was 0, it would
+    // be the case that appending a 0 to a valid list of values would result
+    // in a new valid list. For that reason, Bech32 requires the resulting
+    // checksum to be 1 instead. In Bech32m, this constant was amended.
+    ubyte[] enc = expandHRP(hrp);
+    enc ~= values;
+    uint check = polymod(enc);
+    if (check == encodingConstant(Encoding.Bech32))
+        return Encoding.Bech32;
+    if (check == encodingConstant(Encoding.Bech32m))
+        return Encoding.Bech32m;
+
+    return Encoding.Invalid;
+}
+
+/*******************************************************************************
+
+    Convert from one power-of-2 number base to another
+
+    Params:
+        out_values = values that has converted
+        in_values = values to be converted
+        frombits = a power-of-2 number base of `in_values`
+        tobits = a power-of-2 number base of `out_values`
+        pad = check if the pads are added
+
+    Returns:
+        true if the conversion succeeds
+
+*******************************************************************************/
+
+bool convertBits (ref ubyte[] out_values, const(ubyte)[] in_values,
+    int frombits, int tobits, bool pad) pure nothrow @safe
+{
+    int acc = 0;
+    int bits = 0;
+    const int maxv = (1 << tobits) - 1;
+    const int max_acc = (1 << (frombits + tobits - 1)) - 1;
+    for (size_t i = 0; i < in_values.length; ++i)
+    {
+        int value = in_values[i];
+        acc = ((acc << frombits) | value) & max_acc;
+        bits += frombits;
+        while (bits >= tobits)
+        {
+            bits -= tobits;
+            out_values ~= cast(ubyte)((acc >> bits) & maxv);
+        }
+    }
+    if (pad)
+    {
+        if (bits)
+            out_values ~= cast(ubyte)((acc << (tobits - bits)) & maxv);
+    }
+    else if (bits >= frombits || ((acc << (tobits - bits)) & maxv))
+    {
+        return false;
+    }
+    return true;
+}
+
+// Convert a input string to a lower case string
+version(unittest) char[] toLowerString(const(char)[] input) pure nothrow @safe
+{
+    char[] conv;
+    foreach (c; input)
+        conv ~= toLower(c);
+    return conv;
+}
```

Comment:
There's a function in the std library for this

Reply:
> There's a function in the std library for this

You mean this function, [toLower](https://dlang.org/library/std/uni/to_lower.html#1)?

Reply:
Yes

Reply:
Updated, Ty.